internal/controller/engine/cache.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2024 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+	"strings"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+)
+
+var (
+	_ cache.Cache       = &InformerTrackingCache{}
+	_ TrackingInformers = &InformerTrackingCache{}
+)
+
+// TODO(negz): Propose adding a method like ActiveInformers to the upstream
+// Informers implementation. This would allow us to remove this wrapper.
+
+// An InformerTrackingCache wraps a cache.Cache and keeps track of what GVKs it
+// has started informers for. It takes a blocking lock whenever a new informer
+// is started or stopped, but so does the standard controller-runtime Cache
+// implementation.
+type InformerTrackingCache struct {
+	// The wrapped cache.
+	cache.Cache
+
+	scheme *runtime.Scheme
+
+	mx     sync.RWMutex
+	active map[schema.GroupVersionKind]bool
+}
+
+// TrackInformers wraps the supplied cache, adding a method to query which
+// informers are active.
+func TrackInformers(c cache.Cache, s *runtime.Scheme) *InformerTrackingCache {
+	return &InformerTrackingCache{
+		Cache:  c,
+		scheme: s,
+		active: make(map[schema.GroupVersionKind]bool),
+	}
+}
+
+// ActiveInformers returns the GVKs of the informers believed to currently be
+// active. The InformerTrackingCache considers an informer to become active when
+// a caller calls Get, List, or one of the GetInformer methods. It considers an
+// informer to become inactive when a caller calls the RemoveInformer method.
```

Comment:
This whole file/type just wraps controller runtime's `cache.Informers` type to add this method. Per the comment at the top of this file we could remove it if they added a similar method. They have the data:

https://github.com/kubernetes-sigs/controller-runtime/blob/da68ea45d30c66ec122a7a484e72bd6e3e5ac89a/pkg/cache/internal/informers.go#L109-L113

If we proceed with this direction I'll open an upstream issue to propose adding it.